cli/src/test.rs

```diff
@@ -0,0 +1,280 @@
+use super::error::Result;
+use super::util;
+use ansi_term::Colour;
+use difference::{Changeset, Difference};
+use lazy_static::lazy_static;
+use regex::bytes::{Regex as ByteRegex, RegexBuilder as ByteRegexBuilder};
+use regex::Regex;
+use std::char;
+use std::fs;
+use std::io::{self, Write};
+use std::path::Path;
+use std::str;
+use tree_sitter::{Language, LogType, Parser};
+
+lazy_static! {
+    static ref HEADER_REGEX: ByteRegex = ByteRegexBuilder::new(r"^===+\r?\n([^=]*)\r?\n===+\r?\n")
+        .multi_line(true)
+        .build()
+        .unwrap();
+    static ref DIVIDER_REGEX: ByteRegex = ByteRegexBuilder::new(r"\r?\n---+\r?\n")
+        .multi_line(true)
+        .build()
+        .unwrap();
+    static ref WHITESPACE_REGEX: Regex = Regex::new(r"\s+").unwrap();
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum TestEntry {
+    Group {
+        name: String,
+        children: Vec<TestEntry>,
+    },
+    Example {
+        name: String,
+        input: Vec<u8>,
+        output: String,
+    },
+}
+
+impl Default for TestEntry {
+    fn default() -> Self {
+        TestEntry::Group {
+            name: String::new(),
+            children: Vec::new(),
+        }
+    }
+}
+
+pub fn run_tests_at_path(
+    language: Language,
+    path: &Path,
+    debug: bool,
+    debug_graph: bool,
+    filter: Option<&str>,
+) -> Result<bool> {
+    let test_entry = parse_tests(path)?;
+    let mut _log_session = None;
+    let mut parser = Parser::new();
+    parser.set_language(language)?;
+
+    if debug_graph {
+        _log_session = Some(util::log_graphs(&mut parser, "log.html")?);
+    } else if debug {
+        parser.set_logger(Some(Box::new(|log_type, message| {
+            if log_type == LogType::Lex {
+                io::stderr().write(b"  ").unwrap();
+            }
+            write!(&mut io::stderr(), "{}\n", message).unwrap();
+        })));
+    }
+
+    let mut failures = Vec::new();
+    if let TestEntry::Group { children, .. } = test_entry {
+        for child in children {
+            run_tests(&mut parser, child, filter, 0, &mut failures)?;
+        }
+    }
+
+    if failures.len() > 0 {
+        println!("");
+
+        if failures.len() == 1 {
+            println!("1 failure:")
+        } else {
+            println!("{} failures:", failures.len())
+        }
+
+        print_diff_key();
+        for (i, (name, actual, expected)) in failures.iter().enumerate() {
+            println!("\n  {}. {}:", i + 1, name);
+            print_diff(actual, expected);
+        }
+        Ok(true)
```

Comment:
`Ok(true)` and `Ok(false)` should be switched, or `main.rs` should be modified to exit with 1 on `true`.

Reply:
Thanks @314eter!